sys/net/network_layer/ng_sixlowpan/frag/ng_sixlowpan_frag.c

```diff
@@ -0,0 +1,233 @@
+
+    return local_offset;
+}
+
+static uint16_t _send_nth_fragment(ng_sixlowpan_netif_t *iface, ng_pktsnip_t *pkt,
+                                   size_t payload_len, size_t datagram_size,
+                                   uint16_t offset)
+{
+    ng_pktsnip_t *frag;
+    uint16_t max_frag_size = _floor8(iface->max_frag_size - sizeof(ng_sixlowpan_frag_n_t));
+    uint16_t local_offset = 0, offset_count = 0;
+    ng_sixlowpan_frag_n_t *hdr;
+    uint8_t *data;
+
+    DEBUG("6lo frag: determined max_frag_size = %" PRIu16 "\n", max_frag_size);
+
+    frag = _build_frag_pkt(pkt,
+                           payload_len - offset + sizeof(ng_sixlowpan_frag_n_t),
+                           max_frag_size + sizeof(ng_sixlowpan_frag_n_t));
+
+    if (frag == NULL) {
+        return 0;
+    }
+
+    hdr = frag->next->data;
+    data = (uint8_t *)(hdr + 1);
+
+    hdr->disp_size = byteorder_htons((uint16_t)datagram_size);
+    hdr->disp_size.u8[0] |= NG_SIXLOWPAN_FRAG_N_DISP;
+    hdr->tag = byteorder_htons(_tag);
+    hdr->offset = (uint8_t)(offset >> 3);
+    pkt = pkt->next;    /* don't copy netif header */
+
+    while ((pkt != NULL) || (offset_count == offset)) {   /* go to offset */
+        offset_count += (uint16_t)pkt->size;
+
+        if (offset_count > offset) {    /* we overshot */
+            /* => copy rest of partly send packet snip */
+            uint16_t pkt_offset = offset - (offset_count - ((uint16_t)pkt->size));
+            size_t clen = _min(max_frag_size, pkt->size - pkt_offset);
+
+            memcpy(data, ((uint8_t *)pkt->data) + pkt_offset, clen);
+            local_offset = clen;
+            pkt = pkt->next;
+            break;
+        }
+
+        pkt = pkt->next;
+    }
+
+    if (local_offset < max_frag_size) { /* copy other packet snips */
+        while (pkt != NULL) {
+            size_t clen = _min(max_frag_size - local_offset, pkt->size);
+
+            memcpy(data + local_offset, pkt->data, clen);
+            local_offset += clen;
+
+            if (local_offset == max_frag_size) {
+                break;
+            }
+
+            pkt = pkt->next;
+        }
+    }
+
+    DEBUG("6lo frag: send first fragment (datagram size: %u, "
+          "datagram tag: %" PRIu16 ", offset: %" PRIu8 " (%u bytes), "
+          "fragment size: %" PRIu16 ")\n",
+          (unsigned int)datagram_size, _tag, hdr->offset, hdr->offset << 3,
+          local_offset);
+    ng_netapi_send(iface->pid, frag);
+
+    return local_offset;
+}
+
+void ng_sixlowpan_frag_send(kernel_pid_t pid, ng_pktsnip_t *pkt,
+                            size_t payload_len, size_t datagram_size)
+{
+    ng_sixlowpan_netif_t *iface = ng_sixlowpan_netif_get(pid);
+    uint16_t offset = 0, res;
+
+    if ((res = _send_1st_fragment(iface, pkt, payload_len, datagram_size)) == 0) {
+        /* error sending first fragment */
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    offset += res;
+
+    while (offset < datagram_size) {
```

Comment:
I don't like that this function basically blocks the entire 6lowpan stack until the sending of all fragments is done, but once more: future optimization.